### PR TITLE
feat: multi-provider LLM support (Mistral, OpenAI, Anthropic)

### PR DESCRIPTION
**Pull request by valentinyanakiev**
## Summary
- Replace per-provider LLM adapters with a **unified `LangChainLLMAdapter`** and a **provider factory** — switch between Mistral, OpenAI, and Anthropic by changing `LLM_PROVIDER` env var only
- Add `LLMProvider` enum, provider config fields with Pydantic validation, and backward compatibility for existing `MISTRAL_API_KEY` / `MISTRAL_SMALL_MODEL_NAME` env vars
- Support **local/self-hosted models** (vLLM, Ollama, sglang) via `LLM_BASE_URL`
- Add **per-plugin provider override** via `{PLUGIN_NAME}_LLM_*` env vars
- Harden **structured output JSON parsing** in the guidance plugin (fenced JSON, preamble, trailing text)
- Pass pre-computed embeddings through ChromaDB adapter (fixes onnxruntime dependency issue)
- Add ADR 0005 and full SDD spec artifacts under `specs/002-multi-provider-llm/`

## Key changes

| Area | Files | What |
|------|-------|------|
| Unified adapter | `core/adapters/langchain_llm.py` | Wraps any LangChain `BaseChatModel` with retry + streaming |
| Provider factory | `core/provider_factory.py` | Maps `LLMProvider` → LangChain class with defaults |
| Config | `core/config.py` | `LLMProvider` enum, `LLM_*` fields, validation, backward compat |
| Wiring | `main.py` | Factory integration, per-plugin override, startup logging |
| Structured output | `plugins/guidance/plugin.py` | Robust JSON extraction from varied LLM output formats |
| Removed | `core/adapters/mistral.py`, `core/adapters/openai_llm.py` | Consolidated into unified adapter |

## Test plan
- [x] 60 new tests covering config validation, adapter behavior, factory resolution, and structured output parsing
- [x] 184/184 tests pass (including 4 previously-failing ingest pipeline tests, fixed via conftest mock update)
- [ ] Manual: start engine with `LLM_PROVIDER=openai` and verify valid response
- [ ] Manual: start engine with `LLM_PROVIDER=anthropic` and verify valid response
- [ ] Manual: verify existing `MISTRAL_API_KEY`-only config still works (backward compat)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Multi-provider LLM support (OpenAI, Anthropic, Mistral) with selection via LLM_PROVIDER and unified LangChain-based adapter.
  * Global LLM_* env vars (API key, model, base URL, temperature, max tokens, top-p, timeout) and per-plugin PLUGINNAME_LLM_* overrides.
  * Support for custom/local LLM endpoints via LLM_BASE_URL and startup INFO logging of resolved provider/model.

* **Improvements**
  * Startup validation for LLM config with helpful errors and numeric bounds.
  * More robust structured-output JSON extraction with tolerant parsing.

* **Documentation**
  * Added spec, quickstart, ADR and checklist docs for multi-provider LLM setup.

### CODE REVIEW

File: .env.example [hunk 1]
```diff
@@ -17,10 +17,24 @@ VECTOR_DB_HOST=localhost
...
+# LLM — provider-agnostic configuration
+# Supported providers: mistral, openai, anthropic
+LLM_PROVIDER=mistral
+LLM_API_KEY=
+LLM_MODEL=                     # Default: mistral-large-latest / gpt-4o / claude-sonnet-4-6
+LLM_BASE_URL=                  # Set for local models (e.g., http://localhost:8000/v1)
+LLM_TEMPERATURE=               # 0.0–2.0 (optional)
+LLM_MAX_TOKENS=                # >0 (optional)
+LLM_TOP_P=                     # 0.0–1.0 (optional)
+LLM_TIMEOUT=120                # Seconds
```
[the remark this answers is not in the document]
Comment (valentinyanakiev) [reply .env.example]: Fixed — moved inline comments to separate lines above each assignment.

File: core/adapters/chromadb.py [hunk 2]
```diff
@@ -29,16 +40,31 @@ def __init__(self, host: str, port: int = 8765, credentials: str | None = None)
...
+                collection, embedding_function=None
+            )
+            if query_embeddings is not None:
+                results = col.query(
+                    query_embeddings=query_embeddings, n_results=n_results
+                )
+            else:
+                results = col.query(
+                    query_texts=query_texts, n_results=n_results
+                )
```
[the remark this answers is not in the document]
Comment (valentinyanakiev) [reply core/adapters/chromadb.py]: Already addressed — the code raises `ValueError` when `self._embeddings is None`, so the `query_texts` fallback path no longer exists.

File: core/adapters/chromadb.py [hunk 3]
```diff
@@ -54,10 +80,16 @@ async def ingest(
...
         def _ingest():
-            col = self._client.get_or_create_collection(collection)
-            col.upsert(documents=documents, metadatas=metadatas, ids=ids)
+            col = self._client.get_or_create_collection(
+                collection, embedding_function=None
+            )
+            kwargs: dict = dict(documents=documents, metadatas=metadatas, ids=ids)
+            if embeddings is not None:
+                kwargs["embeddings"] = embeddings
+            col.upsert(**kwargs)
```
[the remark this answers is not in the document]
Comment (valentinyanakiev) [reply core/adapters/chromadb.py]: Already addressed — `ingest()` now raises `ValueError` when `embeddings is None` and calls `col.upsert()` with explicit embeddings.

File: core/adapters/chromadb.py [hunk 4]
```diff
@@ -67,8 +99,11 @@ def _delete():
 
         try:
             await self._retry(_delete)
-        except ValueError:
-            logger.warning("Collection %s not found for deletion", collection)
+        except (ValueError, Exception) as exc:
+            if "not found" in str(exc).lower() or "does not exist" in str(exc).lower():
+                logger.warning("Collection %s not found for deletion, skipping", collection)
+            else:
+                raise
```
[the remark this answers is not in the document]
Comment (valentinyanakiev) [reply core/adapters/chromadb.py]: Fixed — moved the not-found check inside `_delete()` so it catches on first attempt without burning retries.

File: core/config.py [hunk 5]
```diff
@@ -33,12 +43,58 @@ class BaseConfig(BaseSettings):
...
+                self.llm_api_key = self.mistral_api_key
+            if not self.llm_model and self.mistral_model_name:
+                self.llm_model = self.mistral_model_name
+
+        # API key required unless base_url is set (local models)
+        if not self.llm_api_key and not self.llm_base_url:
+            raise ValueError(
+                f"LLM_API_KEY is required for provider '{self.llm_provider.value}'. "
+                "Set LLM_API_KEY or provide LLM_BASE_URL for local models."
+            )
```
[the remark this answers is not in the document]
Comment (valentinyanakiev) [reply core/config.py:76]: Not fixing — this is by design. Each plugin runs in its own container and must have a valid base LLM config (either `LLM_API_KEY`, `MISTRAL_API_KEY` via backward compat, or `LLM_BASE_URL`). Per-plugin overrides (`{PLUGIN}_LLM_*`) are for specialization on top of a working base, not a replacement for all config.

File: pyproject.toml [hunk 6]
```diff
@@ -24,6 +24,7 @@ load-dotenv = "^0.1.0"
 langchain = "^1.1.0"
 langchain-openai = "^1.1.0"
 langchain-mistralai = "^1.1.0"
+langchain-anthropic = "^0.3"
```
[the remark this answers is not in the document]
Comment (valentinyanakiev) [reply pyproject.toml:27]: Not fixing — `^0.3` resolved to `0.3.21` successfully via `poetry lock`. The 0.3.x line has langchain-core requirements compatible with our `langchain ^1.1.0`. Bumping to `^1.0.0` would pull in langchain-core >=1.2.19 which conflicts with our pinned langchain version.

File: specs/002-multi-provider-llm/contracts/provider-config.md [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+
+# K8s deployment for expert plugin
+env:
+  - name: PLUGIN_TYPE
+    value: expert
+  - name: LLM_PROVIDER
+    value: anthropic
+  - name: LLM_API_KEY
+    value: ${ANTHROPIC_KEY}
+```
```
[the remark this answers is not in the document]
Comment (valentinyanakiev) [reply specs/002-multi-provider-llm/contracts/provider-config.md]: Fixed — updated the example to use `GUIDANCE_LLM_PROVIDER` / `EXPERT_LLM_PROVIDER` prefixed env vars matching the implemented interface.

File: tests/core/test_langchain_llm.py [hunk 8]
```diff
@@ -0,0 +1,152 @@
+"""Tests for LangChainLLMAdapter: invoke, stream, retry, message conversion."""
+
+from __future__ import annotations
+
+import asyncio
```
[the remark this answers is not in the document]
Comment (valentinyanakiev) [reply tests/core/test_langchain_llm.py]: Fixed — removed unused `import asyncio`.

File: tests/plugins/test_guidance_structured_output.py [hunk 9]
```diff
@@ -0,0 +1,65 @@
+"""Tests for structured output parsing edge cases in GuidancePlugin."""
+
+from __future__ import annotations
+
+import pytest
```
[the remark this answers is not in the document]
Comment (valentinyanakiev) [reply tests/plugins/test_guidance_structured_output.py]: Fixed — removed unused `import pytest`.
